### PR TITLE
Adding Span<char> IndexOf performance test

### PR DESCRIPTION
**Pull request by ahsonkhan**
cc @shiftylogic, @KrzysztofCwalina, @benaadams 

Please note the (expected?) performance degradation when hardware acceleration is disabled for SpanIndexOfCharAsBytes.
Can we improve the performance for \<byte\> IndexOf when IsHardwareAcceleration == false? Shouldn't it still match the performance of SpanIndexOfChar? Why is it ~2x worst? **Is this expected since we have twice as many bytes to search through compared to the number of chars?**

image

image

image

### CODE REVIEW

**Comment by danmosemsft** [comment]
@stephentoub

**Comment by stephentoub** [comment]
> Is this expected since we have twice as many bytes to search through compared to the number of chars?

You'd need to look at a trace, but I assume that's the reason: it's doing twice as many comparison operations.

File: src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    public class Perf_Span_IndexOf
+    {
+        private const int InnerCount = 100000;
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [InlineData(1)]
+        [InlineData(10)]
+        [InlineData(100)]
+        [InlineData(1000)]
+        [InlineData(10000)]
```
Comment (stephentoub) [inline src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs]: Do we need all of these different sizes?

File: src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        public void SpanIndexOfChar(int size)
+        {
+            Span<char> charSpan = new char[size];
+            charSpan[size/2] = '5';
+
+            int index = 0;
+            foreach (var iteration in Benchmark.Iterations)
```
Comment (stephentoub) [inline src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs]: Nit: var => whatever the type of iteration is

**Comment by ahsonkhan** [comment]
@dotnet-bot test Portable Linux x64 Debug Build

File: src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs (hunk 1 above)
Comment (ahsonkhan) [inline src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs]: Removing some, like 10k is fine. However, it is important to have the 1, 10, 100, to see the overhead cost of vectorization, imo. Will remove the 10k.

**Comment by ahsonkhan** [comment]
> You'd need to look at a trace, but I assume that's the reason: it's doing twice as many comparison operations.

So, this ends up being the scenario:
- IsHardwareAcceleration enabled
   - Don't call Span\<char\>.IndexOf, call Span\<char\>.AsBytes.IndexOf (you will get ~2x faster performance).
- IsHardwareAcceleration disabled
   - Call Span\<char\>.IndexOf because if you call Span\<char\>.AsBytes.IndexOf (you will get ~2x slower performance).

Span\<char\>.IndexOf perf doesn't change regardless since it doesn't leverage vectorization. So, is the recommendation to only call Span\<char\>.AsBytes.IndexOf when you have IsHardwareAcceleration enabled otherwise stick with Span\<char\>.IndexOf?

**Pull request merged**